### PR TITLE
Add recursive createSubDomain and setAddress

### PR DESCRIPTION
**Pull request by VladStarostenko**

### CODE REVIEW

File: waffle-ens/package.json [hunk 1]
```diff
@@ -46,4 +46,4 @@
     "@ensdomains/ens": "^0.4.4",
     "@ensdomains/resolver": "^0.2.4"
   }
-}
+}
```
Comment (marekkirejczyk) [inline waffle-ens/package.json:49]: removing a newline, noise in the repo

File: waffle-ens/src/EnsBuilder.ts [hunk 2]
```diff
@@ -38,33 +42,56 @@ export class ENSBuilder {
...
+    const recursive = options?.recursive || false;
+    const {decodedRootNode, chunks} = getDomainInfo(domain);
+    const isNoTopDomain = (chunks.length > 1);
+    if (!this.registrars[decodedRootNode]) {
+      if (recursive && isNoTopDomain) {
+        await this.createSubDomain(decodedRootNode, options);
+      } else if (recursive && !isNoTopDomain) {
+        await this.createTopLevelDomain(decodedRootNode);
+      } else {
+        throw new Error(
```
Comment (marekkirejczyk) [inline waffle-ens/src/EnsBuilder.ts:67]: extract exception: 
```ts
class MissingDomain extends Error {
  constructor(domain) {
    super(`Domain ${decodedRootNode} doesn't exist.`);
  }
}
throw new InvalidDomain(decodedRootNode);
```

File: waffle-ens/src/EnsBuilder.ts [hunk 3]
```diff
@@ -38,33 +42,56 @@ export class ENSBuilder {
...
+  async setAddress(domain: string, address: string, options?: DomainRegistrationOptions) {
+    const recursive = options?.recursive || false;
+    const {decodedRootNode} = getDomainInfo(domain);
+    if (!this.registrars[decodedRootNode]) {
+      if (recursive) {
+        await this.createSubDomain(decodedRootNode);
+      } else {
+        throw new Error(
+          `Domain ${decodedRootNode} doesn't exist.`
+        );
```
Comment (marekkirejczyk) [inline waffle-ens/src/EnsBuilder.ts:92]: #extractexception

File: waffle-ens/src/EnsBuilder.ts [hunk 4]
```diff
@@ -38,33 +42,56 @@ export class ENSBuilder {
     await this.ens.setSubnodeOwner(HashZero, utils.id(domain), this.registrars[domain].address);
   }
 
-  async createSubDomain(domain: string) {
+  async createSubDomainNotRecursive(domain: string) {
```
Comment (marekkirejczyk) [inline waffle-ens/src/EnsBuilder.ts:45]: `createSubDomainNotRecursive` -> `createSubDomainNonRecursive`
`*Not*` -> `*Non*`
#non

File: waffle-ens/src/EnsBuilder.ts [hunk 5]
```diff
@@ -38,33 +42,56 @@ export class ENSBuilder {
...
+      [domain]: registrar
+    };
   }
 
-  async findRegistrar(rootNode: string) {
-    return this.registrars[rootNode];
+  async createSubDomain(domain: string, options?: DomainRegistrationOptions) {
+    const recursive = options?.recursive || false;
+    const {decodedRootNode, chunks} = getDomainInfo(domain);
+    const isNoTopDomain = (chunks.length > 1);
```
Comment (marekkirejczyk) [inline waffle-ens/src/EnsBuilder.ts:60]: #non

File: waffle-ens/src/EnsBuilder.ts [hunk 6]
```diff
@@ -38,33 +42,56 @@ export class ENSBuilder {
...
+        throw new Error(
+          `Top level domain ${decodedRootNode} doesn't exist.`
+        );
+      }
+    }
+    await this.createSubDomainNotRecursive(domain);
   }
 
-  async setAddress(domain: string, address: string) {
+  async setAddressNotRecursive(domain: string, address: string) {
```
Comment (marekkirejczyk) [inline waffle-ens/src/EnsBuilder.ts:75]: #non

File: waffle-ens/test/EnsBuilder.test.ts [hunk 7]
```diff
@@ -30,34 +30,71 @@ describe('INTEGRATION: Deploy Ens', async () => {
     expect(await ensBuilder.resolver['addr(bytes32)'](namehash('resolver'))).to.eq(ensBuilder.resolver.address);
   });
 
-  it('Create top level domain', async () => {
-    await ensBuilder.createTopLevelDomain('test');
-    expect(await ensBuilder.ens.owner(namehash('test'))).to.eq(ensBuilder.registrars['test'].address);
-  });
+  describe('INTEGRATION: Create domain', async () => {
+    describe('INTEGRATION: Not recursive', async () => {
```
Comment (marekkirejczyk) [inline waffle-ens/test/EnsBuilder.test.ts:34]: #non

File: waffle-ens/test/EnsBuilder.test.ts [hunk 8]
```diff
@@ -30,34 +30,71 @@ describe('INTEGRATION: Deploy Ens', async () => {
...
   });
 
-  it('Set address', async () => {
-    await ensBuilder.setAddress('vlad.dev.ethworks.test', ensBuilder.wallet.address);
-    const node = namehash('vlad.dev.ethworks.test');
-    expect(await ensBuilder.ens.owner(node)).to.eq(ensBuilder.wallet.address);
-    expect(await ensBuilder.resolver['addr(bytes32)'](node)).to.eq(ensBuilder.wallet.address);
-    expect(await ensBuilder.ens.resolver(node)).to.eq(ensBuilder.resolver.address);
+  describe('INTEGRATION: Set address', async () => {
+    describe('INTEGRATION: Not recursive', async () => {
```
Comment (marekkirejczyk) [inline waffle-ens/test/EnsBuilder.test.ts:73]: #non

File: README.md [hunk 9]
```diff
@@ -1 +0,0 @@
-./waffle-cli/README.md
```
Comment (marekkirejczyk) [inline README.md:1]: Add and remove file -> introducing noise in repository

File: waffle-ens/src/utils.ts [hunk 10]
```diff
@@ -21,12 +21,14 @@ interface ENSDomainInfo {
...
+  const isEmptyDomain = (domain === '');
 
-  if (chunks.length === 1 && chunks[0].length > 0) {
+  if (isTopLevelDomain) {
     throw new Error(
       'Invalid domain. Please, enter no top level domain.'
     );
-  } else if (chunks.includes('')) {
+  } else if (isEmptyDomain) {
     throw new Error(
```
Comment (marekkirejczyk) [inline waffle-ens/src/utils.ts:32]: #extractexception InvalidDomain

File: waffle-ens/src/utils.ts [hunk 11]
```diff
@@ -21,12 +21,14 @@ interface ENSDomainInfo {
 
 export const getDomainInfo = (domain: string): ENSDomainInfo => {
   const chunks = domain.split('.');
+  const isTopLevelDomain = (chunks.length === 1 && chunks[0].length > 0);
+  const isEmptyDomain = (domain === '');
 
-  if (chunks.length === 1 && chunks[0].length > 0) {
+  if (isTopLevelDomain) {
     throw new Error(
```
Comment (marekkirejczyk) [inline waffle-ens/src/utils.ts:28]: #extractexception ExpectedSubdomain

File: waffle-ens/src/EnsBuilder.ts [hunk 12]
```diff
@@ -38,33 +42,56 @@ export class ENSBuilder {
...
+  async createSubDomain(domain: string, options?: DomainRegistrationOptions) {
+    const recursive = options?.recursive || false;
+    const {decodedRootNode, chunks} = getDomainInfo(domain);
+    const isNoTopDomain = (chunks.length > 1);
+    if (!this.registrars[decodedRootNode]) {
+      if (recursive && isNoTopDomain) {
+        await this.createSubDomain(decodedRootNode, options);
+      } else if (recursive && !isNoTopDomain) {
+        await this.createTopLevelDomain(decodedRootNode);
+      } else {
```
Comment (marekkirejczyk) [inline waffle-ens/src/EnsBuilder.ts:66]: extract `createDomain`:
```ts
createDomain(...) {
    const isNoTopDomain = (chunks.length > 1);
    if (isNoTopDomain) {
      await this.createSubDomain(decodedRootNode, options);
    } else {
      await this.createTopLevelDomain(decodedRootNode);
    }
}
```

File: waffle-ens/src/EnsBuilder.ts [hunk 13]
```diff
@@ -38,33 +42,56 @@ export class ENSBuilder {
...
     await this.ens.setResolver(node, this.resolver.address);
     await this.resolver['setAddr(bytes32,uint256,bytes)'](node, COIN_TYPE_ETH, address);
   }
+
+  async setAddress(domain: string, address: string, options?: DomainRegistrationOptions) {
+    const recursive = options?.recursive || false;
+    const {decodedRootNode} = getDomainInfo(domain);
+    if (!this.registrars[decodedRootNode]) {
+      if (recursive) {
+        await this.createSubDomain(decodedRootNode);
```
Comment (marekkirejczyk) [inline waffle-ens/src/EnsBuilder.ts:88]: Use extracted `createDomain`

File: waffle-ens/package.json (hunk 1 above)
Comment (marekkirejczyk) [reply waffle-ens/package.json:49]: What about this on?

File: README.md (hunk 9 above)
Comment (marekkirejczyk) [reply README.md:1]: What about this on?

File: waffle-ens/src/errors.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+  }
+}
+
+export class InvalidDomain extends Error {
+  constructor(domain: string) {
+    super(`Invalid domain: '${domain}'`);
+  }
+}
+
+export class NoTopLevelDomain extends Error {
```
Comment (marekkirejczyk) [inline waffle-ens/src/errors.ts:13]: ExpectedTopLevelDomain

**Comment by VladStarostenko** [comment]
Continued in #257

**Pull request closed without merging**